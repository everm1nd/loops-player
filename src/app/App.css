.button {
  font-size: 15px;
  user-select: none;
  border: none;
  text-decoration: none;
}

.button:focus {
  outline: none !important;
}

.button:active {
  opacity: 0.5;
}

.track {
  display: inline-block;
  margin-right: 2px;
}

.track .title {
  margin-bottom: 12px;
  color: #777;
  text-align: center;
}

.clip {
  display: block;
  width: 100px;
  height: 100px;
  background: lightgrey url('components/Clip/play-icon.svg') no-repeat center;
  background-size: 80px;
  cursor: pointer;
  margin-bottom: 2px;
}

@keyframes blinker {
  50% { opacity: 0.5; }
}

.clip.starting, .clip.stopping {
  animation: blinker 1.5s ease-in-out infinite;
}
