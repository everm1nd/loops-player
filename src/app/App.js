import React from 'react';
import 'App.css';
import Clip from "components/Clip";

const App = () => <div className="app">
  <Clip />
</div>

export default App;
