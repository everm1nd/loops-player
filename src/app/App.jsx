import React from 'react';
import 'App.css';
import Track from "components/Track";
import tracksData from "appData";
import { initPlaybackState, clickPlaybackState, tickPlaybackState } from 'playbackStateTransformer'
import Tone from 'tone'

const QUANTIZATION = "2m"
const BPM = 110;

class App extends React.Component {
  constructor(props) {
    super(props)
    this.state = initPlaybackState(tracksData)
    this._handleClipClick = this._handleClipClick.bind(this)
    this._tick = this._tick.bind(this)
  }

  componentDidMount() {
    Tone.Transport.bpm.value = BPM;
    Tone.Transport.scheduleRepeat(this._tick, QUANTIZATION);
  }

  _tick(time) {
    console.log(Tone.Transport.seconds.toFixed(2))
    this.setState(
      tickPlaybackState,
      this._stopTransportUnlessPlaying
    )
  }

  _allClipsStopped() {
    return this.state.tracks.every((track) => (
      track.clips.every(clip => clip.playbackState === "stopped")
    ))
  }

  _startTransport() {
    if (Tone.Transport.state === "stopped") Tone.Transport.start()
  }

  _stopTransportUnlessPlaying() {
    if (this._allClipsStopped()) {
      Tone.Transport.stop()
      Tone.Transport.cancel()
    }
  }

  _handleClipClick(trackId, clipId) {
    this.setState(
      clickPlaybackState(this.state, { trackId, clipId }),
      this._startTransport
    )
  }

  renderTracks() {
    return this.state.tracks.map( (track, index) =>
      <Track id={index} key={track.name} {...track} onClick={this._handleClipClick} />
    )
  }

  render() {
    return <div className="app">{this.renderTracks()}</div>
  }
}

export default App;
