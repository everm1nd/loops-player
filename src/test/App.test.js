import React from 'react';
import { shallow } from 'enzyme';

import togglePlayback from 'playbackStateMachine'
togglePlayback.onClick = jest.fn().mockReturnValue('magicstate')

import App from 'App';
import Tone from 'tone';

describe('App', () => {
  const wrapper = shallow(<App />);

  it('renders an app', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('._handleClipClick', () => {
    it('assigns state from playbackStateMachine to clip state', () => {
      wrapper.instance()._handleClipClick(0,0)
      expect(wrapper.state().tracks[0].clips[0].playbackState).toEqual('magicstate')
    })
  })

  describe('._tick', () => {
    it('changes state of staring clips to started', () => {
      const state = wrapper.state()
      state.tracks[0].clips[0].playbackState = "starting"
      wrapper.setState(state)

      wrapper.instance()._tick(0)
      expect(wrapper.state().tracks[0].clips[0].playbackState).toEqual("started")
    });

    it('changes state of stopping clips to stopped', () => {
      const state = wrapper.state()
      state.tracks[0].clips[0].playbackState = "stopping"
      wrapper.setState(state)

      wrapper.instance()._tick(0)
      expect(wrapper.state().tracks[0].clips[0].playbackState).toEqual("stopped")
    });
  });

  describe('componentDidMount', () => {
    beforeEach(() => {
      Tone.Transport.scheduleRepeat.mockReset();
      const wrapper = shallow(<App />);
    })

    it('sets a scheduleRepeat loop on global transport', () => {
      expect(Tone.Transport.scheduleRepeat).toHaveBeenCalled();
    })
  });
})
