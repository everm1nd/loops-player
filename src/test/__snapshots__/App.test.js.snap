// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders an app 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Track
          clips={
                    Array [
                              Object {
                                "id": 0,
                                "name": "Funky Clip 1",
                                "playbackState": "stopped",
                                "url": "/media/drums-1.wav",
                              },
                              Object {
                                "id": 1,
                                "name": "Funky Clip 2",
                                "playbackState": "stopped",
                                "url": "/media/drums-2.wav",
                              },
                              Object {
                                "id": 2,
                                "name": "Funky Clip 3",
                                "playbackState": "stopped",
                                "url": "/media/drums-1.wav",
                              },
                            ]
          }
          color="blue"
          id={0}
          name="Drums"
          onClick={[Function]}
/>,
        <Track
          clips={
                    Array [
                              Object {
                                "id": 0,
                                "name": "Soft Clip 1",
                                "playbackState": "stopped",
                                "url": "/media/melody-1.wav",
                              },
                              Object {
                                "id": 1,
                                "name": "Soft Clip 2",
                                "playbackState": "stopped",
                                "url": "/media/melody-1.wav",
                              },
                              Object {
                                "id": 2,
                                "name": "Soft Clip 3",
                                "playbackState": "stopped",
                                "url": "/media/melody-1.wav",
                              },
                            ]
          }
          color="deeppink"
          id={1}
          name="Melody"
          onClick={[Function]}
/>,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "Drums",
        "nodeType": "function",
        "props": Object {
          "clips": Array [
            Object {
              "id": 0,
              "name": "Funky Clip 1",
              "playbackState": "stopped",
              "url": "/media/drums-1.wav",
            },
            Object {
              "id": 1,
              "name": "Funky Clip 2",
              "playbackState": "stopped",
              "url": "/media/drums-2.wav",
            },
            Object {
              "id": 2,
              "name": "Funky Clip 3",
              "playbackState": "stopped",
              "url": "/media/drums-1.wav",
            },
          ],
          "color": "blue",
          "id": 0,
          "name": "Drums",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Melody",
        "nodeType": "function",
        "props": Object {
          "clips": Array [
            Object {
              "id": 0,
              "name": "Soft Clip 1",
              "playbackState": "stopped",
              "url": "/media/melody-1.wav",
            },
            Object {
              "id": 1,
              "name": "Soft Clip 2",
              "playbackState": "stopped",
              "url": "/media/melody-1.wav",
            },
            Object {
              "id": 2,
              "name": "Soft Clip 3",
              "playbackState": "stopped",
              "url": "/media/melody-1.wav",
            },
          ],
          "color": "deeppink",
          "id": 1,
          "name": "Melody",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Track
            clips={
                        Array [
                                    Object {
                                      "id": 0,
                                      "name": "Funky Clip 1",
                                      "playbackState": "stopped",
                                      "url": "/media/drums-1.wav",
                                    },
                                    Object {
                                      "id": 1,
                                      "name": "Funky Clip 2",
                                      "playbackState": "stopped",
                                      "url": "/media/drums-2.wav",
                                    },
                                    Object {
                                      "id": 2,
                                      "name": "Funky Clip 3",
                                      "playbackState": "stopped",
                                      "url": "/media/drums-1.wav",
                                    },
                                  ]
            }
            color="blue"
            id={0}
            name="Drums"
            onClick={[Function]}
/>,
          <Track
            clips={
                        Array [
                                    Object {
                                      "id": 0,
                                      "name": "Soft Clip 1",
                                      "playbackState": "stopped",
                                      "url": "/media/melody-1.wav",
                                    },
                                    Object {
                                      "id": 1,
                                      "name": "Soft Clip 2",
                                      "playbackState": "stopped",
                                      "url": "/media/melody-1.wav",
                                    },
                                    Object {
                                      "id": 2,
                                      "name": "Soft Clip 3",
                                      "playbackState": "stopped",
                                      "url": "/media/melody-1.wav",
                                    },
                                  ]
            }
            color="deeppink"
            id={1}
            name="Melody"
            onClick={[Function]}
/>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "Drums",
          "nodeType": "function",
          "props": Object {
            "clips": Array [
              Object {
                "id": 0,
                "name": "Funky Clip 1",
                "playbackState": "stopped",
                "url": "/media/drums-1.wav",
              },
              Object {
                "id": 1,
                "name": "Funky Clip 2",
                "playbackState": "stopped",
                "url": "/media/drums-2.wav",
              },
              Object {
                "id": 2,
                "name": "Funky Clip 3",
                "playbackState": "stopped",
                "url": "/media/drums-1.wav",
              },
            ],
            "color": "blue",
            "id": 0,
            "name": "Drums",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Melody",
          "nodeType": "function",
          "props": Object {
            "clips": Array [
              Object {
                "id": 0,
                "name": "Soft Clip 1",
                "playbackState": "stopped",
                "url": "/media/melody-1.wav",
              },
              Object {
                "id": 1,
                "name": "Soft Clip 2",
                "playbackState": "stopped",
                "url": "/media/melody-1.wav",
              },
              Object {
                "id": 2,
                "name": "Soft Clip 3",
                "playbackState": "stopped",
                "url": "/media/melody-1.wav",
              },
            ],
            "color": "deeppink",
            "id": 1,
            "name": "Melody",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
